hwloc/topology-rsmi.c

```diff
@@ -0,0 +1,268 @@
+  enum hwloc_type_filter_e filter;
+  rsmi_status_t ret;
+  unsigned nb, i;
+
+  assert(dstatus->phase == HWLOC_DISC_PHASE_IO);
+
+  hwloc_topology_get_type_filter(topology, HWLOC_OBJ_OS_DEVICE, &filter);
+  if (filter == HWLOC_TYPE_FILTER_KEEP_NONE)
+    return 0;
+
+  rsmi_init(0);
+
+  ret = rsmi_num_monitor_devices(&nb);
+  if (RSMI_STATUS_SUCCESS != ret || !nb) {
+    if (RSMI_STATUS_SUCCESS != ret)
+      fprintf(stderr, "Failed to get device count\n");
+    rsmi_shut_down();
+    return 0;
+  }
+
+  for (i=0; i<nb; i++) {
+    rsmiPciInfo_t pci;
+    hwloc_obj_t osdev, parent;
+    char buffer[64];
+
+    osdev = hwloc_alloc_setup_object(topology, HWLOC_OBJ_OS_DEVICE, HWLOC_UNKNOWN_INDEX);
+    snprintf(buffer, sizeof(buffer), "rsmi%u", i);
+    osdev->name = strdup(buffer);
+    osdev->depth = HWLOC_TYPE_DEPTH_UNKNOWN;
+    osdev->attr->osdev.type = HWLOC_OBJ_OSDEV_GPU;
+
+    hwloc_obj_add_info(osdev, "Backend", "RSMI");
+    hwloc_obj_add_info(osdev, "GPUVendor", "Advanced Micro Devices");
+
+    buffer[0] = '\0';
+    get_device_name(i, buffer, sizeof(buffer));
+    hwloc_obj_add_info(osdev, "GPUModel", buffer);
+
+    buffer[0] = '\0';
+    get_device_serial_number(i, buffer, sizeof(buffer));
+    if (buffer[0] != '\0')
+      hwloc_obj_add_info(osdev, "AMDSerial", buffer);
+
+    buffer[0] = '\0';
+    get_device_unique_id(i, buffer);
+    if (buffer[0] != '\0')
+      hwloc_obj_add_info(osdev, "AMDUUID", buffer);
+
+    parent = NULL;
+    get_device_pci_info(i, &pci);
+    parent = hwloc_pci_find_parent_by_busid(topology, pci.domain, pci.bus, pci.device, 0);
+    get_device_pci_linkspeed(i, &(parent->attr->pcidev.linkspeed));
+    if (!parent)
+      parent = hwloc_get_root_obj(topology);
+
+    hwloc_insert_object_by_parent(topology, parent, osdev);
+  }
+
+  rsmi_shut_down();
+  return 0;
+}
+
+static struct hwloc_backend *
+hwloc_rsmi_component_instantiate(struct hwloc_topology *topology,
+				 struct hwloc_disc_component *component,
+				 unsigned excluded_phases __hwloc_attribute_unused,
+				 const void *_data1 __hwloc_attribute_unused,
+				 const void *_data2 __hwloc_attribute_unused,
+				 const void *_data3 __hwloc_attribute_unused)
+{
+  struct hwloc_backend *backend;
+
+  backend = hwloc_backend_alloc(topology, component);
+  if (!backend)
+    return NULL;
+  backend->discover = hwloc_rsmi_discover;
+  return backend;
+}
+
+static struct hwloc_disc_component hwloc_rsmi_disc_component = {
+  "rsmi",
+  HWLOC_DISC_PHASE_IO,
+  HWLOC_DISC_PHASE_GLOBAL,
+  hwloc_rsmi_component_instantiate,
+  5, /* after pci, and after cuda since likely less useful */
+  1,
+  NULL
+};
+
+static int
+hwloc_rsmi_component_init(unsigned long flags)
+{
+  if (flags)
+    return -1;
+  if (hwloc_plugin_check_namespace("rsmi", "hwloc_backend_alloc") < 0)
+    return -1;
+  if (!dlopen("librocm_smi64.so", RTLD_NOW | RTLD_GLOBAL)) {
+    fprintf(stderr, "RSMI configured, but wasn't found.\n");
+    return -1;
+  }
```

Comment:
Why do you need this dlopen? The rsmi hwloc plugin won't be loaded unless this librocm_smi64.so is available since it will depend on it.

Reply:
Let see hwloc is build in another machine. But rocm_smi isn't installed in the machine hwloc is running, we stop here instead crash when with init.

Reply:
If topology-rsmi.c is built statically inside libhwloc (not as plugin), libhwloc.so will depend on librocm_smi64.so explicitly, hence libhwloc.so will fail to load, and the program will fail to start because of the missing librocm64_smi.so
If built as a plugin, libhwloc.so will work fine, the program will start. When hwloc will load plugins, libhwloc_rsmi.so will fail to load because it needs librocm_smi64.so. That's actually the whole point of plugins: build everything but don't require all dependencies to be available at runtime, only those available will have their hwloc plugins loaded.
It'll work fine without dlopen, we already do that for multiple other dependencies.

Reply:
Ok